src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -163,13 +166,33 @@ internal Cci.IFieldReference GetModuleVersionId(Cci.ITypeReference mvidType)
         {
             if (_mvidField == null)
             {
+                Debug.Assert(!IsFrozen);
                 Interlocked.CompareExchange(ref _mvidField, new ModuleVersionIdField(this, mvidType), null);
             }
 
             Debug.Assert(_mvidField.Type.Equals(mvidType));
             return _mvidField;
         }
 
+        internal Cci.IFieldReference GetOrAddInstrumentationPayloadRoot(int analysisKind, Cci.ITypeReference payloadRootType)
+        {
+            InstrumentationPayloadRootField payloadRootField;
+            if (!_instrumentationPayloadRootFields.TryGetValue(analysisKind, out payloadRootField))
+            {
+                Debug.Assert(!IsFrozen);
+                payloadRootField = _instrumentationPayloadRootFields.GetOrAdd(analysisKind, (kind) => new InstrumentationPayloadRootField(this, kind, payloadRootType));
```

Comment:
For what it's worth, the other case like this used a lambda. The reason I prefer the lambda here is that it guarantees that creating the field object will be created exactly once, which might matter if the constructor were some. This code will execute once per analysis kind, so the overhead of creating the lambda will never be significant.